### PR TITLE
add transform-enforcing compartment wrapper

### PR DESCRIPTION
**Pull request by warner**
This is still a work in process, but I'd love to get some feedback on the approach from @erights and @kriskowal . Also, I don't really understand how `new Foo()` works, I'd appreciate some help making it callable as `new newCompartment()` instead of just `newCompartment()`.

refs #377

### CODE REVIEW

**Comment by warner** [comment]
Ok, ready for review. Questions:

* does this belong in SES-shim? (I think  yes, by putting it in the SES-shim monorepo, it stands a better chance of keeping up-to-date with changes to the Compartment constructor)
* is the package name reasonable?
* is the API name (`wrapInescapableCompartment`) reasonable?
* can you think of any other sneaky ways to escape the "inescapable" that I should explore+test?

I don't have a test that will probably fail when @kriskowal's changes land (to give each compartment a distinct `Compartment` constructor).. I don't think I can write such a test until after there are differences to be observed. But this will need to be rewritten when that happens, otherwise child/grandchild compartments will get the parent's constructor and will get the wrong module map.

**Comment by warner** [comment]
maybe the package ought to be named `@agoric/inescapable-compartment-wrapper` or something..

**Comment by warner** [comment]
Oh, also I must add the right `package.json` magic to make this importable under `-r esm` before it lands, I need that from the agoric-sdk side..

**Comment by warner** [comment]
`@agoric/inescapable-compartment-wrapper`? `@agoric/inescapable-compartment-transforms`? `@agoric/wrap-compartment-inescapably`?

`import { wrapInescapableCompartment } from X`? `import { wrapCompartment } from X`? `import { wrapCompartmentInescapably } from X`? `import { wrap } from X`?

**Comment by kriskowal** [comment]
> `@agoric/inescapable-compartment-wrapper`? `@agoric/inescapable-compartment-transforms`? `@agoric/wrap-compartment-inescapably`?
> 
> `import { wrapInescapableCompartment } from X`? `import { wrapCompartment } from X`? `import { wrapCompartmentInescapably } from X`? `import { wrap } from X`?

* inescapable-compartment
* jail
* enforcer

Getting a little dark with inescapability metaphors. What else can’t you leave?

* garden
* event-horizon
* border
* boundary

Maybe a step back

* recompartment
* partition

**Comment by kriskowal** [comment]
If you’re worried about prematurely publishing, is it possible to keep this close to the code that uses it until it’s firmer? Given that we don’t know how broadly applicable this pattern is beyond metering, I’d like to keep it close to the meter.

**Comment by kriskowal** [comment]
There are also `peerDependencies` but they are odd. Hat tip to Domenic Denicola for that contribution to the npm ecosystem.

File: packages/compartment-wrapper/compartment-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    // which will cease to be the case soon (their module tables are
+    // different). TODO: update this when that happens, we need something
+    // like c.globalThis.Compartment = wrap(c.globalThis.Compartment), but
+    // there are details to work out.
+    c.globalThis.Compartment = NewCompartment;
+
+    return c;
+  };
+
+  // ensure `isinstance(c, Compartment)` still holds true
```
Comment (kriskowal) [inline packages/compartment-wrapper/compartment-wrapper.js:57]: I see what you did here, `c instanceof Compartment` being the JS spelling of Python’s `isinstance(c, Compartment)`. 😉

File: packages/compartment-wrapper/package.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  "devDependencies": {
+    "eslint": "^6.8.0",
+    "eslint-config-airbnb-base": "^14.0.0",
+    "eslint-config-prettier": "^6.9.0",
+    "eslint-plugin-eslint-comments": "^3.1.2",
+    "eslint-plugin-import": "^2.19.1",
+    "eslint-plugin-prettier": "^3.1.2",
+    "@rollup/plugin-node-resolve": "^6.1.0",
+    "rollup-plugin-terser": "^5.1.3",
+    "ses": "^0.9.0+1-dev",
```
Comment (kriskowal) [inline packages/compartment-wrapper/package.json:34]: Should be `^0.9.1` I think. Would not hurt to also have a `peerDependency`.

**Comment by warner** [comment]
I guess the high-order bit is whether this lives in SES-shim or agoric-sdk, but anticipating it breaking when you introduce per-compartment `Compartment` constructors, I'm inclined to think it belongs in SES-shim. It's certainly not agoric-sdk -specific.

But yeah, I could move it into an un-separately-published module within `@agoric/import-bundle`, which is the first customer. The plus side is not needing to choose a public name yet, the downside is that it'll stop working (poorly) when you make that change. "Poorly" because I don't think this code has a direct dependency on any particular version of SES (it doesn't import SES, it merely expects to be given a `Compartment` argument that behaves in a particular way), and when each compartment has a distinct constructor, the wrapper will start confusing those together, silently. I could make `@agoric/import-bundle` depend upon a particular version of SES, but it really needs to be the same version as what `@agoric/install-ses` uses or things will get *really* weird.

It kind of comes down to how we anticipate the SES ecosystem going. We expect that there will be helper libraries that are widely used but live outside the Compartment proposal. We don't know what else will be in these libraries yet, nor what they ought to be called. So yeah, maybe waiting until we have a second use case is a good idea.

So I'll ask for feedback on the implementation here, but I won't land this PR. Instead, I'll move the module into agoric-sdk `@agoric/import-bundle`. I'll add a note to try and remind us that when we change `@agoric/install-ses` to use a newer version of SES (with the constructor-per-compartment), we need to rewrite the wrapper too.

File: packages/compartment-wrapper/compartment-wrapper.js (hunk 1 above)
Comment (warner) [reply packages/compartment-wrapper/compartment-wrapper.js:57]: 20 years of python leaves some marks :)

File: packages/compartment-wrapper/package.json (hunk 2 above)
Comment (warner) [reply packages/compartment-wrapper/package.json:34]: Ooh, yes, that is the right answer. I'll add that `peerDependency` to `@agoric/import-bundle` where this code will live.

**Comment by warner** [comment]
closed in favor of putting the wrapper in agoric-sdk for now

**Pull request closed without merging**